### PR TITLE
android-tools: init at 31.0.0p1

### PR DESCRIPTION
**Pull request by primeos**
~~TODO:
[ 52%] Built target fipsmodule
[ 52%] Generating chacha/chacha-x86_64.S
[ 52%] Generating test/trampoline-x86_64.S
[ 52%] Generating cipher_extra/aes128gcmsiv-x86_64.S
[ 52%] Generating cipher_extra/chacha20_poly1305_x86_64.S
[ 52%] Generating err_data.c
go: golang.org/x/crypto@v0.0.0-20200622213623-75b288015ac9: Get "https://proxy.golang.org/golang.org/x/crypto/@v/v0.0.0-20200622213623-75b288015ac9.mod": dial tcp: lookup proxy.golang.org on [::1]:53: read udp [::1]:56418->[::1]:53: read: connection refused
make[2]: *** [vendor/boringssl/crypto/CMakeFiles/crypto.dir/build.make:104: vendor/boringssl/crypto/err_data.c] Error 1
make[2]: *** Deleting file 'vendor/boringssl/crypto/err_data.c'
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:1351: vendor/boringssl/crypto/CMakeFiles/crypto.dir/all] Error 2
make: *** [Makefile:171: all] Error 2
builder for '/nix/store/yib15nh8kzgrf94ijbgsjmbppcs59jij-android-tools-31.0.0p1.drv' failed with exit code 2
error: build of '/nix/store/yib15nh8kzgrf94ijbgsjmbppcs59jij-android-tools-31.0.0p1.drv' failed~~

###### Motivation for this change

See #75603 and https://github.com/NixOS/nixpkgs/issues/75603#issuecomment-851010621. This let's us finally build the Android tools from source.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by primeos** [comment]
The good news is that only the vendored BoringSSL seems to use Go modules (`./vendor/boringssl/go.sum`) but unfortunately we seem to lack an equivalent of #112500 for Go. We could use a workaround but since this'll likely affect more and more Nixpkgs packages it would be nice if we could implement a general abstraction. I'll open a dedicated issue for that.

**Comment by musfay** [comment]
simg2img, img2simg and append2simg binaries are already provided by simg2img package.

**Comment by primeos** [comment]
@musfay good to know. We should probably still use the bundled ones to avoid version conflicts (if they're used by the other tools) but then we should at least use `lowPrio` to avoid conflicts with the `simg2img` package or move those binaries out of `bin/`.

File: pkgs/tools/misc/android-tools/default.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
+{ lib, stdenv, fetchurl
+, cmake, perl, go
+, protobuf, zlib, gtest, brotli, lz4, zstd, libusb1, pcre2
+}:
+
+stdenv.mkDerivation rec {
+  pname = "android-tools";
+  version = "31.0.0p1";
+
+  src = fetchurl {
```
Comment (IvarWithoutBones) [inline pkgs/tools/misc/android-tools/default.nix:10]: Could this be using `fetchFromGitHub`?

**Review by IvarWithoutBones — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/android-tools/default.nix (hunk 1 above)
Comment (musfay) [reply pkgs/tools/misc/android-tools/default.nix:10]: We need to use release tarball as the repo only have build-system and documentation. Upstream doesn't vendorize platform-tools source code in repo. They use CPack to create tarball from only used source files. 
https://github.com/nmeum/android-tools/tree/master/vendor

Comment (Atemu) [reply pkgs/tools/misc/android-tools/default.nix:10]: How many sources would we need to declare in Nix to achieve the same? I'd prefer if all sources were declared transparently instead of relying on an opaque tarball.

Comment (musfay) [reply pkgs/tools/misc/android-tools/default.nix:10]: Well in that case we need to checkout some of AOSP repositories. Those repositories are actually HUGE and Google doesn't provide snapshots from tags. Checking out repositories takes nearly 1 hour on my computer.

Comment (Atemu) [reply pkgs/tools/misc/android-tools/default.nix:10]: The whole tree is obviously huge but we shouldn't need all of it. That's why I was asking.

Comment (musfay) [reply pkgs/tools/misc/android-tools/default.nix:10]: We can declare the source files but there is 13-14 repos and lots of files all aroud the tree. Leaving the tracking job to CPack would be much easier. I don't agree tarballs are opaque as you can reproduce them. We can create our tarballs or vendorize CPack output in a repository.

Comment (musfay) [reply pkgs/tools/misc/android-tools/default.nix:10]: Is it legally OK if we mirror some of Google's repositories on other platforms like [this one](https://github.com/musfay/adb).

**Comment by primeos** [comment]
I haven't done any extensive testing yet but while implementing a workaround for fetching the Go modules I've noticed that setting `GOFLAGS = [ "-mod=vendor" ];` is enough and that the Go modules are apparently not required (at least the build succeeds without them).

File: lib/licenses.nix [hunk 2]
```diff
@@ -760,6 +760,11 @@ lib.mapAttrs (n: v: v // { shortName = n; }) ({
     # channel and NixOS images.
   };
 
+  unicode-dfs-2015 = spdx {
```
Comment (primeos) [inline lib/licenses.nix:763]: Not completely sure if adding this here is worth it but given that `unicode-dfs-2016` is already there it might make sense.

File: pkgs/tools/misc/android-tools/default.nix (hunk 1 above)
Comment (Atemu) [reply pkgs/tools/misc/android-tools/default.nix:10]: > there is 13-14 repos

Yeah, that's a bit much.

> I don't agree tarballs are opaque as you can reproduce them. 

The problem with that is that the reproduction is not transparent and in this case probably not even reproducible at a different point in time.  
You don't know anything about how CPack was made nor which version of the sources it fetched.    
Google's repo doesn't have a single "revision" or an equivalent; it's an aggregation of whatever happens to be at the top of a few hundred repos at a certain point in time.      
It'd only become transparent if the generation was declared in Nix because then you'd have to lock the exact revision and hash of every included repo.

As I said though, too many repos. I was hoping for, like, 5 which would've been a feasible amount to maintain ourselves.

> Is it legally OK if we mirror some of Google's repositories

The APL2 allows you to redistribute source code in any way you like but, why?

Comment (musfay) [reply pkgs/tools/misc/android-tools/default.nix:10]: > > I don't agree tarballs are opaque as you can reproduce them.
> 
> The problem with that is that the reproduction is not transparent and in this case probably not even reproducible at a different point in time.
> You don't know anything about how CPack was made nor which version of the sources it fetched.
> Google's repo doesn't have a single "revision" or an equivalent; it's an aggregation of whatever happens to be at the top of a few hundred repos at a certain point in time.
> It'd only become transparent if the generation was declared in Nix because then you'd have to lock the exact revision and hash of every included repo.

It makes sense now. Thanks for the definition!

> As I said though, too many repos. I was hoping for, like, 5 which would've been a feasible amount to maintain ourselves.
Yep there is lots of repos. 
> > Is it legally OK if we mirror some of Google's repositories
> 
> The APL2 allows you to redistribute source code in any way you like but, why?
Just asked, offtopic.

**Comment by musfay** [comment]
```
failed to initialize build cache at /homeless-shelter/.cache/go-build: mkdir /homeless-shelter: permission denied
make[2]: *** [vendor/boringssl/crypto/CMakeFiles/crypto.dir/build.make:104: vendor/boringssl/crypto/err_data.c] Error 1
make[2]: *** Deleting file 'vendor/boringssl/crypto/err_data.c'
make[1]: *** [CMakeFiles/Makefile2:1351: vendor/boringssl/crypto/CMakeFiles/crypto.dir/all] Error 2
make: *** [Makefile:171: all] Error 2
builder for '/nix/store/7zfz1im7k03y87hb051mg7b67hr5irrn-android-tools-31.0.0p1.drv' failed with exit code 2
error: build of '/nix/store/7zfz1im7k03y87hb051mg7b67hr5irrn-android-tools-31.0.0p1.drv' failed
```
Oops.

**Comment by Atemu** [comment]
You'll probably have to set some Go environment variable to use /build for the build instead of $HOME.

**Comment by musfay** [comment]
Ofborg fails too. We can create adb-boringssl package with required revision of boringssl as it's only required by adb. Using nixpkgs's boringssl also fails.

**Comment by primeos** [comment]
Thanks, it looks like the Nix build sandboxing on my non-NixOS build host isn't properly working (anymore?).
I've built it on a NixOS host now.

**Comment by musfay** [comment]
I can confirm that everything works as expected. Flashed some images via fastboot. Tested adb through WiFi, pushed some files, had no issues.

**Comment by primeos** [comment]
@musfay thanks for the testing! I've also used `adb` today and didn't notice any issues.

I'll merge this now (it works, we finally have a FLOSS package for adb and fastboot, and we can still improve the Nixpkgs expressions later).

**Comment by Atemu** [comment]
Should we update programs.adb to use adb from here instead?

**Comment by primeos** [comment]
Sounds like a good idea to me (cc @Mic92).
Interestingly `androidenv.androidPkgs_9_0.platform-tools` doesn't have `meta.license` set which seems like a bug...

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -1039,6 +1039,10 @@ in
 
   analog = callPackage ../tools/admin/analog {};
 
+  android-tools = lowPrio (callPackage ../tools/misc/android-tools {
+    stdenv = if stdenv.targetPlatform.isAarch64 then gcc10Stdenv else stdenv;
```
Comment (Atemu) [inline pkgs/top-level/all-packages.nix:1043]: Why is this needed?

Comment (primeos) [reply pkgs/top-level/all-packages.nix:1043]: Was likely required for https://github.com/NixOS/nixpkgs/pull/124992#issuecomment-852043128 ;)
